lighthouse-core/audits/third-party-summary.js

```diff
@@ -57,41 +85,98 @@ class ThirdPartySummary extends Audit {
-      const entity = thirdPartyWeb.getEntity(attributeableURL);
-      if (!entity) continue;
+      const attributableURL = BootupTime.getAttributableURLForTask(task, jsURLs);
 
-      const entityStats = entities.get(entity) || {...defaultEntityStat};
+      const urlStat = byURL.get(attributableURL) || {...defaultStat};
       const taskDuration = task.selfTime * cpuMultiplier;
       // The amount of time spent on main thread is the sum of all durations.
-      entityStats.mainThreadTime += taskDuration;
+      urlStat.mainThreadTime += taskDuration;
       // The amount of time spent *blocking* on main thread is the sum of all time longer than 50ms.
       // Note that this is not totally equivalent to the TBT definition since it fails to account for FCP,
       // but a majority of third-party work occurs after FCP and should yield largely similar numbers.
-      entityStats.blockingTime += Math.max(taskDuration - 50, 0);
-      entities.set(entity, entityStats);
+      urlStat.blockingTime += Math.max(taskDuration - 50, 0);
+      byURL.set(attributableURL, urlStat);
+    }
+
+    // Map each URL's stat to a particular third party entity.
+    /** @type {Map<ThirdPartyEntity, string[]>} */ const urls = new Map();
+    for (const [url, urlStat] of byURL.entries()) {
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity) {
+        byURL.delete(url);
+        continue;
+      }
+      const entityStat = byEntity.get(entity) || {...defaultStat};
+      entityStat.transferSize += urlStat.transferSize;
+      entityStat.mainThreadTime += urlStat.mainThreadTime;
+      entityStat.blockingTime += urlStat.blockingTime;
+      byEntity.set(entity, entityStat);
+
+      const entityURLs = urls.get(entity) || [];
+      entityURLs.push(url);
+      urls.set(entity, entityURLs);
     }
 
-    return entities;
+    return {byURL, byEntity, urls};
+  }
+
+  /**
+   * @param {ThirdPartyEntity} entity
+   * @param {{byEntity: Map<ThirdPartyEntity, Summary>, byURL: Map<string, Summary>, urls: Map<ThirdPartyEntity, string[]>}} summaries
+   * @param {Summary} stats
+   * @return {Array<URLSummary>}
+   */
+  static getSubItems(entity, summaries, stats) {
+    const entityURLs = summaries.urls.get(entity) || [];
+    let items = entityURLs
+      .map(url => /** @type {URLSummary} */ ({url, ...summaries.byURL.get(url)}))
+      // Filter out any cases where byURL was missing entries.
+      .filter((stat) => stat.transferSize > 0)
+      // Sort by blocking time first, then transfer size to break ties.
+      .sort((a, b) => (b.blockingTime - a.blockingTime) || (b.transferSize - a.transferSize));
+
+    const runningSummary = {transferSize: 0, blockingTime: 0};
+    const minTransferSize = Math.max(MIN_TRANSFER_SIZE_FOR_SUBITEMS, stats.transferSize / 20);
+    const maxSubItems = Math.min(MAX_SUBITEMS, items.length);
+    let i = 0;
+    while (i < maxSubItems &&
```

Comment:
can I sell you on just making this a for loop and checking if `runningSummary.transferSize === 0` instead of `i === 0`? :)

Reply:
Sold! but also added a `blockingTime` check